### PR TITLE
Enhanced Access Control: Removal of Global Permissions, Direct Permission Setting, and External Contract Calls by IP Owners

### PR DESCRIPTION
**Pull request by kingster-will**
This PR includes three significant changes to the access control mechanism in our Solidity contracts, specifically in the `AccessController.sol` file.

1. **Removal of Global Permissions:** Removes the concept of global permissions. Previously, permissions were able to set at a global level. With this change, permissions are now can only set at a more granular level, providing better control and security.

2. **Direct Permission Setting by IP Owners:** Allows IP owners to directly set permissions, no need through the `IPAccount` contract. This change simplifies the process of setting permissions and makes it more intuitive for IP owners. 

3. **IP Owners Can Call External Contracts:** Allows IP owners to call any external contracts through their `IPAccount`. This provides IP owners with greater flexibility and control over their IP, enabling them to interact with a wider range of contracts and services.

### CODE REVIEW

File: contracts/access/AccessController.sol [hunk 1]
```diff
@@ -164,21 +146,22 @@ contract AccessController is IAccessController, AccessManagedUpgradeable, UUPSUp
...
-            !IModuleRegistry($.moduleRegistry).isRegistered(signer)
-        ) {
-            revert Errors.AccessController__BothCallerAndRecipientAreNotRegisteredModule(signer, to);
-        }
         // Must be a valid IPAccount
         if (!IIPAccountRegistry($.ipAccountRegistry).isIpAccount(ipAccount)) {
             revert Errors.AccessController__IPAccountIsNotValid(ipAccount);
         }
-        // Owner can call all functions of all modules
+        // Owner can call any contracts either registered module or unregistered/external contracts
```
Comment (jdubpark) [inline contracts/access/AccessController.sol:153]: Need to revisit any risk of opening up IPAccounts to call external contracts freely (previously it was through "external" modules).

Is the "pass-through" risk discussed some time ago still relevant in this context? cc @Spablob @Ramarti

**Review by jdubpark — APPROVED** [review]

File: contracts/access/AccessController.sol [hunk 2]
```diff
@@ -141,8 +123,8 @@ contract AccessController is IAccessController, AccessManagedUpgradeable, UUPSUp
         if (permission > 2) {
             revert Errors.AccessController__PermissionIsNotValid();
         }
-        if (!IModuleRegistry($.moduleRegistry).isRegistered(msg.sender) && ipAccount != msg.sender) {
-            revert Errors.AccessController__CallerIsNotIPAccount();
+        if (ipAccount != msg.sender && IIPAccount(payable(ipAccount)).owner() != msg.sender) {
```
Comment (Ramarti) [inline contracts/access/AccessController.sol:126]: I think we could remove the 1st parameter of `setPermission` (`ipAccount`) in favor of msg.sender, and here you check that sender is an ipAccount or the owner

**Review by LeoHChen — APPROVED** [review]

File: contracts/access/AccessController.sol (hunk 1 above)
Comment (kingster-will) [reply contracts/access/AccessController.sol:153]: It is only the IP owner can call external contracts.  no change to the behavior of "pass-through."

Comment (Ramarti) [reply contracts/access/AccessController.sol:153]: That comment states that the owner can call anything through the IPA, which is the expected outcome of ERC6551

File: contracts/access/AccessController.sol (hunk 2 above)
Comment (kingster-will) [reply contracts/access/AccessController.sol:126]: iiuc, Do you mean remove the `ipAccount` from the `setPermission()` parameters? 
Line 112-114 is for checking whether `ipAccount` is address(0).
As `ipAccount` is a field of permission, we need to store it in the permission mapping; hence, I don't feel we could remove it.

File: contracts/access/AccessController.sol [hunk 3]
```diff
@@ -164,21 +146,22 @@ contract AccessController is IAccessController, AccessManagedUpgradeable, UUPSUp
...
         if (!IIPAccountRegistry($.ipAccountRegistry).isIpAccount(ipAccount)) {
             revert Errors.AccessController__IPAccountIsNotValid(ipAccount);
         }
-        // Owner can call all functions of all modules
+        // Owner can call any contracts either registered module or unregistered/external contracts
         if (IIPAccount(payable(ipAccount)).owner() == signer) {
             return;
         }
+        if (
+            to != address(this) &&
```
Comment (Ramarti) [inline contracts/access/AccessController.sol:158]: "If we are calling to an address that is not the IPAccount, and such destination address is not a registered module, and the signer is not a registered module", revert.

I'm not sure what are we preventing here. I'm going to try using reversed logic, please @kingster-will correct me if I'm wrong

- If the destination we are calling is the IPAccount address -> OK
or
- destination is a registered module -> OK
or
- signer is a registered module -> OK

OK here means that we are going to check if there are granular or module level permissions in the lines below.

In which cases the signer would a registered module?

File: contracts/access/AccessController.sol (hunk 1 above)
Comment (kingster-will) [reply contracts/access/AccessController.sol:153]: right, ERC6551 doesn't restrict target of execute() of IPAccount.

File: contracts/access/AccessController.sol (hunk 3 above)
Comment (jdubpark) [reply contracts/access/AccessController.sol:158]: I believe the first check is checking if "to" is Access Controller, not IP Account.

Comment (kingster-will) [reply contracts/access/AccessController.sol:158]: I see.
`to != address(this)` is for checking it whether call to AccessController but IPAccount.
It verifies whether either the signer or the to is a registered module.

File: contracts/access/AccessController.sol (hunk 2 above)
Comment (Ramarti) [reply contracts/access/AccessController.sol:126]: Yeah, reviewing it, the thing is that this method can be called by the ipAccount or the owner. I read it as only callable by IPAccount. Nevermind!

File: contracts/access/AccessController.sol (hunk 3 above)
Comment (LeoHChen) [reply contracts/access/AccessController.sol:158]: please add proper comments in the code @kingster-will

Comment (kingster-will) [reply contracts/access/AccessController.sol:159]: Added comments to clarify it.

**Review by LeoHChen — APPROVED** [review]

**Pull request merged**